### PR TITLE
Fix #39: Redis won't start with ipv6 disabled

### LINKED ISSUE #39: `redis-server` installation fails when IPv6 is disabled
I used this Redis role after OS hardening role which disabled all IPv6 settings in `sysctl.conf`. That resulted in the failure of `apt-get install redis-server` (I'm on Ubuntu server) command.

Error output is:
```
Reading package lists...
Building dependency tree...
Reading state information...
redis-server is already the newest version (5:4.0.9-1ubuntu0.1).
0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
1 not fully installed or removed.
After this operation, 0 B of additional disk space will be used.
Setting up redis-server (5:4.0.9-1ubuntu0.1) ...
Job for redis-server.service failed because a timeout was exceeded.
See \"systemctl status redis-server.service\" and \"journalctl -xe\" for details.
invoke-rc.d: initscript redis-server, action \"start\" failed.
* redis-server.service - Advanced key-value store
   Loaded: loaded (/lib/systemd/system/redis-server.service; disabled; vendor preset: enabled)
   Active: activating (auto-restart) (Result: timeout) since Wed 2019-01-30 15:26:36 MSK; 7ms ago
     Docs: http://redis.io/documentation,
           man:redis-server(1)

Jan 30 15:26:36 eps-lw-banking systemd[1]: \u001b[0;1;31m\u001b[0;1;39m\u001b[0;1;31mFailed to start Advanced key-value store.\u001b[0m
dpkg: error processing package redis-server (--configure):
 installed redis-server package post-installation script subprocess returned error exit status 1
Errors were encountered while processing:
 redis-server
```

The problem is that default `redis.conf` from `redis-server` package contains `bind 127.0.0.1 ::1` line, so Redis tries to bind on disabled IPv6 address and hangs.

I found discussions of this problem on [Ubuntu](https://bugs.launchpad.net/ubuntu/+source/redis/+bug/1752207) and [Debian](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=900284) bugtrackers. I understand that this is more `redis-server` package problem than the problem of this role, but it would be great if this role had a workaround for this situation.

By default, this role is configured to set `bind` to `127.0.0.1` (`redis_bind_interface` setting), so I found the following solution: I changed the order of Redis configuration and Redis installation tasks in the play, and I also added the task for creating the `/etc/redis` dir:
```
- name: Ensure Redis configuration dir exists.
  file:
    path: "{{ redis_conf_path | dirname }}"
    state: directory
    mode: 0755

- name: Ensure Redis is configured.
  template:
    src: redis.conf.j2
    dest: "{{ redis_conf_path }}"
    mode: 0644
  notify: restart redis

# Setup/install tasks.
- include_tasks: setup-Debian.yml
  when: ansible_os_family == 'Debian'
```

Now, when `redis-server` installation happens, the installer uses preconfigured `redis.conf` and all works even if IPv6 is totally disabled.

### PR DESCRIPTION
**Pull request by agrrh**
Fixes #39 by putting config in place before installing a package

### CODE REVIEW

**Comment by agrrh** [comment]
Should be fixed now %)

**Comment by agrrh** [comment]
@geerlingguy Could you take a look at this, please?

**Comment by joelpittet** [comment]
@agrrh I'm running into this but got it to work by just changing one line using whatever Ubuntu comes with
`vagrant_box: geerlingguy/drupal-vm` I think `18.04`

I'm trying out your patch to see if it works as a solution but wondering if your OS also needs this?
`sed -i 's/^supervised no/supervised systemd/' /etc/redis/redis.conf`
https://github.com/geerlingguy/drupal-vm/issues/2107

**Comment by joelpittet** [comment]
No dice, I destroyed and built with this patch applied and didn't get any better results.

**Comment by joelpittet** [comment]
I'm experimenting with this patch + `supervised auto` but wondering why the permission changes @agrrh ?

**Comment by joelpittet** [comment]
No luck still running into this timeout, I even stopped configuring and letting the package just install with defaults.

**Comment by agrrh** [comment]
Hey there, @joelpittet !

My case was all around IPv6 disabled and Redis tried to bind to ::1

`supervised` parameter configures how Redis handles various signals and ENV variables, so it wasn't the thing.

Permission changes are related to difference in Ubuntu/CentOS permissions for `/etc` contents, I guess. Sadly, there's a confusing message in commit and right now I can't remember particular cause.

**Comment by joelpittet** [comment]
No luck still running into this timeout after install status check, I even stopped "configuring" and letting the package just install with defaults `/etc/redis/redis.conf`.

@agrrh Thanks for letting me know the details, I can see how this patch could help you and I'm +1 to committing it.

File: tasks/main.yml [hunk 1]
```diff
@@ -8,6 +8,19 @@
...
 
+- name: Ensure Redis configuration dir exists.
+  file:
+    path: "{{ redis_conf_path | dirname }}"
+    state: directory
+    mode: 0755
+
+- name: Ensure Redis is configured.
+  template:
+    src: redis.conf.j2
```
Comment (geerlingguy) [inline tasks/main.yml:19]: Wouldn't putting this in place prior to Redis' installation cause Redis to overwrite the file?

**Comment by agrrh** [comment]
Nope, it works as expected.

Dpkg leaves existing files alone according to docs:
https://www.debian.org/doc/debian-policy/ap-pkg-conffiles.html

Same logic for rpm-based distros:
https://docs.fedoraproject.org/ro/Fedora_Draft_Documentation/0.1/html/RPM_Guide/ch09s05s03.html

вт, 17 нояб. 2020 г., 21:33 Jeff Geerling <notifications@github.com>:

File: tasks/main.yml (hunk 1 above)
Comment (agrrh) [reply tasks/main.yml:19]: Nope, it works as expected.

Dpkg leaves existing files alone according to docs:
https://www.debian.org/doc/debian-policy/ap-pkg-conffiles.html

Same logic for rpm-based distros:
https://docs.fedoraproject.org/ro/Fedora_Draft_Documentation/0.1/html/RPM_Guide/ch09s05s03.html

Comment (agrrh) [reply tasks/main.yml:19]: And same for Arch:
https://wiki.archlinux.org/index.php/Pacman#What_happens_during_package_install/upgrade/removal

Comment (joelpittet) [reply tasks/main.yml:19]: @geerlingguy I can confirm that Redis keeps the existing config that was in place before the install.

**Pull request merged**